pkg/persistentvolumeclaim/v1alpha1/pvc.go

```diff
@@ -0,0 +1,124 @@
+package v1alpha1
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+// pvc holds the api's pvc objects
+type pvc struct {
+	object *v1.PersistentVolumeClaim
+}
+
+// PVCList holds the list of pvc instances
+type PVCList struct {
+	Items []*pvc
+}
+
+// listBuilder enables building an instance of
+// pvclist
+type listBuilder struct {
+	list    *PVCList
+	filters predicateList
+}
+
+// WithAPIList builds the list of pvc
+// instances based on the provided
+// pvc list api instance
+func (b *listBuilder) WithAPIList(pvcs *v1.PersistentVolumeClaimList) *listBuilder {
+	if pvcs == nil {
+		return b
+	}
+	for _, p := range pvcs.Items {
+		b.list.Items = append(b.list.Items, &pvc{&p})
+	}
+	return b
+}
+
+// WithAPIList builds the list of pvc
+// instances based on the provided
+// pvc's api instances
+func (b *listBuilder) WithAPIObject(pvcs ...v1.PersistentVolumeClaim) *listBuilder {
+	if len(pvcs) == 0 {
+		return b
+	}
+	for _, p := range pvcs {
+		b.list.Items = append(b.list.Items, &pvc{&p})
+	}
+	return b
+}
+
+// List returns the list of pvc
+// instances that was built by this
+// builder
+func (b *listBuilder) List() *PVCList {
+	if b.filters == nil || len(b.filters) == 0 {
+		return b.list
+	}
+	filtered := ListBuilder().List()
+	for _, pvc := range b.list.Items {
+		if b.filters.all(pvc) {
+			filtered.Items = append(filtered.Items, pvc)
+		}
+	}
+	return filtered
+}
+
+// List returns the API list of pvc
+// instances that was built by this
+// builder
+func (b *listBuilder) APIList() *v1.PersistentVolumeClaimList {
+	plist := &v1.PersistentVolumeClaimList{}
+	for _, pvc := range b.list.Items {
+		plist.Items = append(plist.Items, *pvc.object)
+	}
+	return plist
+}
+
+// ListBuilder returns a instance of listBuilder
+func ListBuilder() *listBuilder {
+	return &listBuilder{list: &PVCList{Items: []*pvc{}}}
+}
+
+// predicate defines an abstraction
+// to determine conditional checks
+// against the provided csp instance
+type predicate func(*pvc) bool
+
+// IsBound is a predicate to filter out pvcs
+// which is bounded
+func IsBound() predicate {
```

Comment:
We can have a new predicate that check for nil.
```go
func IsNil() predicate {
  return func(p *pvc) bool {
    return p.IsNil()
  }
}

func (p *pvc) IsNil() bool {
  return p.object == nil
}
```

Reply:
added